### PR TITLE
REFACTOR-#6362: `__setitem__` on multiple columns should be evaluated lazily

### LINKED ISSUE #6362: PERF: `__setitem__` on multiple columns should be evaluated lazily
NOTE: I haven't done any clear benchmarking for this.

In the following reproducer:
```python
import modin.pandas as pd
import pandas as vpd
import numpy as np

cols = [f"feature_{i}" for i in range(4)]
cols.append('labels')
df = pd.read_csv('data.txt', header=None)
df.columns = cols
val_df = df.sample(frac=0.2)
train_df = df.drop(val_df.index)

train_means = train_df[cols[:-1]].mean()
train_std = train_df[cols[:-1]].std()
train_df[cols[:-1]] = (train_df[cols[:-1]]- train_means)/train_std
```

data: [data.txt](https://github.com/modin-project/modin/files/11969394/data.txt)

`__setitem__` on multiple columns should not materialize partitions immediately.

### PR DESCRIPTION
**Pull request by pyrito**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #6362 
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/utils.py [hunk 1]
```diff
@@ -345,6 +340,28 @@ def broadcast_item(
...
+            return np.broadcast_to(item, to_shape)
+    except ValueError:
+        from_shape = np.array(item).shape
+        raise ValueError(
+            f"could not broadcast input array from shape {from_shape} into shape "
+            + f"{to_shape}"
+        )
+
+
+def _maybe_reindex_item(obj, row_lookup, col_lookup, item, need_columns_reindex=True):
```
Comment (mvashishtha) [inline modin/pandas/utils.py:359]: Can we remove the "maybe" and make this method only valid for series and dataframes?

File: modin/pandas/dataframe.py [hunk 2]
```diff
@@ -2451,13 +2452,22 @@ def __setitem__(self, key, value):
...
-                )
+
+                # If we have a modin Series or DataFrame, let's not convert it to
+                # a numpy array if we can help it.
+                if isinstance(value, (DataFrame, Series)):
+                    item = _maybe_reindex_item(
+                        self, slice(None), key, value, need_columns_reindex=False
+                    )
+                else:
+                    item = broadcast_item(
```
Comment (mvashishtha) [inline modin/pandas/dataframe.py:2463]: So `_maybe_reindex_item` gets called in either code path here (either directly here, or in `broadcast_item`)? Can we always call it directly here?

**Review by mvashishtha — COMMENTED** [review]
@pyrito have you checked whether the PandasOnPython and BaseOnPython query compilers can handle inserting a dataframe instead of a numpy array?

I also left some comments about the implementation details.

File: modin/pandas/dataframe.py (hunk 2 above)
Comment (pyrito) [reply modin/pandas/dataframe.py:2463]: Yeah actually this was dumb thinking on my part. I could just return early in `broadcast_item`.

**Review by mvashishtha — COMMENTED** [review]
this solution generally looks ok. I left some comments about making it neater

File: modin/pandas/utils.py [hunk 3]
```diff
@@ -358,6 +348,24 @@ def broadcast_item(
         # New value for columns/index make that reindex add NaN values
         if axes_to_reindex:
             item = item.reindex(**axes_to_reindex)
+
+    # Allow the query_compiler to dictate what to do instead of materializing the data
+    if isinstance(item, (DataFrame, Series)):
```
[the remark this answers is not in the document]
Comment (pyrito) [reply modin/pandas/utils.py:353]: @mvashishtha technically we are still doing a broadcast in the sense that we are reshaping the dataframe, it's just that we may not always be returning a numpy array so I think only the docs really need to be updated.

**Comment by anmyachev** [comment]
@dchigarev so we can close it, right?

**Comment by anmyachev** [comment]
Looks like something similar already done in #6618. Feel free to reopen if needed.

**Pull request closed without merging**